Xpanes: convert to connected nodeboxes.

sofar opened the pull request:
I've rewritten this to use connected nodeboxes, but with a caveat. In
order to make flat nodes look better, I'm keeping one non-connected
pane that is flat around to convert flat sections to the flat nodes
instead of connected, as these look better and are easier to work
with. Once more sides are needed we convert the panes on the fly to
connected nodes and recalculate the shape.

We don't paint any of the half-panes that the previous generation of
xpanes did. There's no need and it's harder to work with. Updating the
nodes also seems more natural and placement and removal works straight
forward.

This is *missing* any compat/conversion code, so it's not yet complete.
I expect to use an LBM to perform the conversion, which is the right
tool to do so.

Check out https://youtu.be/13M6Hz8oQ1I for how it looks.

- [x] - conversion code.

Rui914 commented:
```
local nodenames = {}
for i = 1, 15 do
	nodenames[#nodenames + 1] = "xpanes:" .. name .. "_" .. i
end

minetest.register_lbm({
	name = "xpanes:convert_xpanes_to_connected_nodeboxes_" .. name,
	nodenames = nodenames ,
	action = function(pos)
		minetest.set_node(pos, {name = "xpanes:" .. name"})
		for i = 0, 3 do
			local dir = minetest.facedir_to_dir(i)
			update_pane(vector.add(pos, dir))
		end
	end
})
```

Rui914 commented on mods/xpanes/init.lua:
```diff
@@ -1,156 +1,148 @@
+	local groups = table.copy(def.groups)
+	groups.pane = 1
+	groups.not_in_creative_inventory = 1
+	minetest.register_node("xpanes:" .. name, {
```
"xpanes:" > ":xpanes:"

sofar replied:
thanks, I'll get that.

kilbith commented:
Good job. 2 remarks : 

- The current crossed-panes (the node you place at first without connection) is oftenly used for various purposes like table legs or thing for suspending lighting blocks. Your initial node is just a simple pane.
- At [1:55](https://youtu.be/13M6Hz8oQ1I?t=115) of your video, the selection box is very small and hard to point when the pane is crossed.

But that shouldn't block this PR as the benefit is very interesting (15 nodes less for each pane).

sofar commented:
@Rui914 thanks, that indeed looks right, although I wonder if we can make it a single lbm that is used for all derivatives.

sofar commented:
@kilbith the selection box being small is something I think I can fix in core.

Fixer-007 commented:
Nice! I will look at them.

Rui914 commented:
```lua
local nodenames = {}

function xpanes.register_pane(name, def)
	for i = 1, 15 do
		nodenames[#nodenames + 1] = "xpanes:" .. name .. "_" .. i
	end
	...
end

minetest.after(0, function()
	minetest.register_lbm({
		name = "xpanes:replace_legacy_panes",
		nodenames = nodenames,
		action = function(pos, node)
			local n = node.name:match("xpanes:.+_(%d+)")
			node.name = node.name:sub(1, -(#n + 1))
			minetest.set_node(pos, node)
		end
	})
end)
```
Too hackish and operation is unconfirmed.

sofar commented:
I have a better idea, give me a sec.

sofar commented:
@rui914's code was close, but no cigar :)

Problem is that lbm's can't run in ```minetest.after(0``` context, which is correct.

However, we can easily use aliases to tell core that all the old nodes are actually flat panes and belong to ```group:pane```, and then the lbm becomes trivial, we just call ```update_pane()``` on all panes and around all panes.

Tested conversion from master to xpanes several times, good results.

sofar commented:
Oh and this conversion will also work excellent for mods using the ```register_pane()``` API.

kilbith commented:
@paramat Would it be wise to mark it as milestone ? As it's a feature and the freeze happens tomorrow...

sofar commented:
I don't consider this WIP anymore. Especially with the selection boxes fixed, this is a nice reduction of nodedef count, and the code isn't any more complex.

sofar commented:
I'm taking a closer look at the texture usage in this patch. It could be improved.

sofar commented:
Fixed up the texture usage and docs.

Ferk commented:
Did this not make it for the 0.4.14?

ShadowNinja commented on mods/xpanes/init.lua:
```diff
@@ -1,156 +1,147 @@
+			connect_back = {{-1/31, -1/2, 1/31, 1/31, 1/2, 1/2}},
+			connect_right = {{1/31, -1/2, -1/31, 1/2, 1/2, 1/31}},
+		},
+		connects_to = {"group:pane", "group:stone", "group:glass", "group:wood", "group:tree"},
```
This should actually connect to anything walkable, and I think that type of connection is pretty common, so it should have core support.

Ferk replied:
Imho, it should be possible to override it in the definition. Using something like this:

     connects_to = def.connects_to or {"group:pane", "group:stone", "group:glass", "group:wood", "group:tree"}

paramat replied:
I guess since the connects_to list has to be a list of specified nodes it has to be kept reasonably short and it's not possible to check for walkable.
A settable 'connects_to' could be good though?

ShadowNinja commented:
I don't like how you've got a hybrid now where server-side updates are used sometimes and client-side ones other times.  A `connect_none` or similar option should probably be added for this type of use case.

sofar commented:
We don't implement facedir rotation for connected nodeboxes, so I felt this would become a kludge quickly. Yes, it's one extra node, but the benefit is that we never have half panes, which I think is visually much more appealing.

ShadowNinja commented:
Whether or not half-panes are desirable is debatable...

sofar commented:
The flat pane itself makes it "connect" to stuff that's not in the explicit connect_to list. That means that I can place it in between two furnaces and it still looks connected:

![](http://i.imgur.com/QV74IKG.jpg)

You can't do this with the half-panes.

ShadowNinja commented:
@sofar: But then it breaks if you try to use more than one pane between the nodes.  That's unexpected behavior.

sofar commented:
@shadowninja, that's not what happens, for example:

![](http://i.imgur.com/98izn7w.jpg)

Ferk commented:
@sofar I think he means in the X axis, not Y.

imho, the main issue is that we lose the + shaped node that some people used for tables and decorations, like kilbith said. So some maps or schematics might be a bit broken with this.

The lack of facedir rotation for a "+" node wouldn't be a problem. ShadowNinja's idea of a connect_none property looks interesting. It would simplify it a lot and you'll still get it to connect to stuff whenever it's actually supposed to connect_to.

sofar commented:
> I think he means in the X axis, not Y.

? it works the same in all axes, X, Z. I just stacked them to make the screenshot smaller, but Y is irrelevant.

Ferk commented:
When spreading the panes in parallel there's necessarily gotta be a difference between connected_to nodes or not. Thought on second read, I doubt that's what ShadowNinja meant... my bad

![](http://i.imgur.com/CkfcfkK.png?1)

lightonflux commented:
Will this be merged before the freeze?

sofar commented:
No rebase needed, but it needs 1-2 devs to actually use this code a bit.

The behavior is very different, but I think it's intuitive for players and understandable, and will lead to easier building than the old panes.

@nore, @paramat, others, please try this in-game.

norey commented:
@sofar why are you tagging me? How am I related to this project?

sofar commented:
@norey apologies, someone else is using several handles of which one is one letter off from yours, so github autocompleted it to yours. My bad.

norey commented:
@sofar don't worry about it

sofar commented:
Needs review - if this is OK then I can make more time addressing issues if they are found, but there are people who want to add more panes who are blocked/have conflicts with this PR.

Fixer-007 commented:
What I will miss a lot are glass half-panes :/ that can create nice effect of glass entrance.

paramat commented:
Man, i need to review ths.

paramat commented:
@rubenwardy if possible can you look at this? Seems good work and deserves attention. I'll try to review also.

paramat commented:
Watched the video and the behaviour looked good.
I won't miss the half-panes, the loss of the cross shape might annoy some, but the state of xpanes is not good and i feel a significant rework is welcome.
Current xpanes update 'on construct' which i guess is not called for placement by voxelmanip or schematic? I guess xpanes will not rotate if a schematic is rotated?

paramat commented on mods/xpanes/init.lua:
```diff
@@ -1,156 +1,147 @@
 	end
-	if sum == 0 then
-		sum = 15
+	local asidenode = minetest.get_node(aside)
```
Seems unused in this function?

sofar replied:
Oh IC, yes, this was needed before I optimized this code to use `find_nodes_in_area()`. I'll remove it.

paramat commented:
Reviewed and code seems ok. Also added a line comment and replied to a previous line cmment.
Overall i like this.

paramat commented on mods/xpanes/init.lua:
```diff
@@ -1,156 +1,147 @@
+		sounds = def.sounds,
+		node_box = {
+			type = "connected",
+			fixed = {{-1/31, -1/2, -1/31, 1/31, 1/2, 1/31}},
```
Just wondering why 1/31? To avoid z-fighting with 1/32?

sofar replied:
The pane ends up 1/16 (1 px) thick this way, and centered in the node, which has the best visual.

paramat replied:
I must be misunderstanding something, the flat pane and the current xpanes use 1/32.
2 * 1/32 = 1/16.

paramat replied:
1/2 - 1/32 != 1/31 ?

sofar replied:
Yes, the current xpanes pane is smaller than 1px. I changed this on purpose to 1px thickness, which feels more coherent.

paramat commented:
Otherwise +1.

paramat commented on mods/xpanes/init.lua:
```diff
@@ -1,156 +1,146 @@
+			type = "connected",
+			fixed = {{-1/31, -1/2, -1/31, 1/31, 1/2, 1/31}},
+			connect_front = {{-1/31, -1/2, -1/2, 1/31, 1/2, -1/31}},
+			connect_left = {{-1/2, -1/2, -1/31, -1/31, 1/2, 1/31, 1/31}},
```
Has 7 entries.

t4im replied:
Also these double nested tables `{{}}` are unnecessary and use additional persistent ram. Only a little of course, but they'll just end up copied everywhere.

sofar replied:
oh, this needs fixing.

sofar replied:
turns out the first 6 elements were all OK, so this wouldn't have caused any problems. Removed it otherwise, of course.

paramat commented:
^ Just saw an error.

paramat commented:
The build fail issue is minor.

paramat commented:
@rubenwardy since this s a major change it would be good to hear your opinion too.

sofar commented:
Removed that erroneous extra array element.

rubenwardy commented on mods/xpanes/init.lua:
```diff
@@ -1,156 +1,146 @@
-	node = node or minetest.get_node(pos)
-	local name = node.name
-	if not name or node.name:sub(1, 7) ~= "xpanes:" then
+local function swap(pos, node, name, param2)
```
this seems confusingly names to me, it's more of a set_node which avoids excess work

paramat replied:
Seems reasonable to me.

paramat commented:
I think we must be misunderstanding each other.

The current xpanes correctly use 1/32 and 2 * 1/32 = 1/16 = 1 pixel thick:
```
local half_boxes = {
-	{0,     -0.5, -1/32, 0.5,  0.5, 1/32},
-	{-1/32, -0.5, 0,     1/32, 0.5, 0.5},
-	{-0.5,  -0.5, -1/32, 0,    0.5, 1/32},
-	{-1/32, -0.5, -0.5,  1/32, 0.5, 0}
-}
-
-local full_boxes = {
-	{-0.5,  -0.5, -1/32, 0.5,  0.5, 1/32},
-	{-1/32, -0.5, -0.5,  1/32, 0.5, 0.5}
-}
```
Youe flat pane correctly uses 1/32:
```
+		node_box = {
+			type = "fixed",
+			fixed = {{-1/2, -1/2, -1/32, 1/2, 1/2, 1/32}},
+		},
```
However your connected nodebox variant uses 1/31 which is incorrect:
```
		node_box = {
			type = "connected",
			fixed = {{-1/31, -1/2, -1/31, 1/31, 1/2, 1/31}},
			connect_front = {{-1/31, -1/2, -1/2, 1/31, 1/2, -1/31}},
			connect_left = {{-1/2, -1/2, -1/31, -1/31, 1/2, 1/31}},
			connect_back = {{-1/31, -1/2, 1/31, 1/31, 1/2, 1/2}},
			connect_right = {{1/31, -1/2, -1/31, 1/2, 1/2, 1/31}},
},
```
2 * 1/31 != 1/16 these are slightly thinner panes.

So i can't approve this PR yet.

sofar commented:
oh HAHAHAHA, 31, not 32. Yeah, I totally read *over* that :D

paramat commented:
Cool will +1 once corrected.

sofar commented:
<lego batman> *first try*</lego batman>

paramat commented:
:+1:

t4im commented on mods/xpanes/init.lua:
```diff
@@ -1,156 +1,146 @@
+		connect_sides = { "left", "right" },
+	})
+
+	local groups = table.copy(def.groups)
```
`groups` is already defined, thus the linter warning.
You could reassign, rename or use two [`do-end` blocks](http://www.lua.org/pil/4.2.html) to solve this.

sofar replied:
False positive: the linter is designed to watch out for "potentially" bad pieces of code, but in this case the reassignment is on purpose.

Fixing false positive static analysis warnings just to fix a warning is somewhat bad practice: the tools are designed to give false positives, forcing people to fix false positives is bad form, and the output of false positives should always be carefully inspected.

So, I'd like to leave it as is.

Ferk replied:
The problem is not the reassignment, but the re-declaration. Just remove the "local".

    groups = table.copy(def.groups)

sofar replied:
My point is that I prefer keeping the `local` here.

Re-using variable names and declaring them again with `local` makes it explicit that I'm creating a new, unused version.

t4im replied:
The linter is pedantic as cobble here, but not actually throwing you a false positive.

You are defining an already defined variable in the same chunk without making it syntactically explicit, that you wanted the previous scope to end.
It's such syntactic discrepancies that the linter is meant to show you, whether they lead to a bug or not.

> Re-using variable names and declaring them again with local makes it explicit that I'm creating a new, unused version.

Lua's `do-end` syntax allows you that. It makes the variable scopes explicit instead of hiding them.

```lua
do
   local var = "why is my inventory full of cobble?"
   print(var)
end -- var isn't defined after this point anymore
```

Maybe you don't want the linter to be that strict to everyone, which is your prerogative, and an understandable decision, but even then leaving it as-is isn't an option either. You'll have to reconfigure the linter to not warn about this particular issue anymore.

Or this will just let every PR fail after this is merged, which is the only reason i'm even pointing this out. It could kill the linter off the project, before it actually warned about something more serious.

sofar replied:
Oh FCOL already.

t4im replied:
> Oh FCOL already.

Did not intend to frustrate you with it.
Rather to prevent the following PR-failings.

```lua
ignore = {"411"} -- `411  │ Redefining a local variable. `
```
Is still an option, as pointed out.

sofar replied:
I personally thought the linter was a great addition, I'm more frustrated that we're spending more time discussing it than coding ;).

t4im replied:
Heh, fair point.

sofar commented:
Updated to avoid static analysis warning.

paramat commented:
Nice solution.

paramat commented:
This is not essential but you could add yourself to README code credits and license.txt code license.
Rubenwardy seems ok with this PR so it can probably be merged soon.

paramat commented:
It would be good to add yourself to README.txt code credits and to license.txt code license, but this could be done in a later commit if you prefer.
Rubenwardy seems fairly happy with this so it could be merged soon.

paramat commented:
Would you like me to merge this and add credits myself, if you're short on time?

sofar commented:
@paramat I didn't really care two days back, but I just looked at `git blame init.lua` and noticed that I've basically redid the entire thing :)

I've put myself in both files (`README.txt` and `license.txt`) underneath the old authors, so, thanks for asking ;)

paramat commented:
Thanks, after the massive job i did on updating credits and licensing we really need to keep on top of that stuff now.

paramat commented:
54235f32766167bbd55cae5715d6e01754715ce6
